torch/_inductor/codegen/triton_foreach.py

```diff
@@ -0,0 +1,294 @@
+import itertools
+from typing import List
+
+import sympy
+
+from .. import config
+from ..ir import Layout
+from ..utils import ceildiv, sympy_product
+from ..virtualized import V
+from .common import IndentedBuffer, Kernel
+from .triton_overrides import TritonOverrides
+from .triton_utils import (
+    config_of,
+    IterationRangesRoot,
+    signature_of,
+    TritonCSEVariable,
+    TritonPrinter,
+)
+
+texpr = TritonPrinter().doprint
+
+
+class ListTracker:
+    def __init__(self, var, arg_names, layouts):
+        self.var: TritonCSEVariable = var
+        self.arg_names: List[str] = arg_names
+        self.layouts: List[Layout] = layouts
+        self.indexers = [layout.make_indexer() for layout in layouts]
+        self.range_trees: List[IterationRangesRoot] = []
+        self.index_vars: List[sympy.symbol] = []
+        name = "xindex"
+        for layout in layouts:
+            tree = IterationRangesRoot(
+                "xindex", sympy_product(layout.size), name[0], 0, V.kernel
+            )
+            self.range_trees.append(tree)
+            self.index_vars.append(tree.construct(layout.size))
+
+    def codegen_tile_ptrs(self, code: IndentedBuffer, list_index: int):
+        index_expr = self.index_expr(list_index)
+
+        nodes = [
+            V.kernel.range_tree_nodes[v]
+            for v in sorted(index_expr.free_symbols, key=lambda s: s.name)
+        ]
+        for node in nodes:
+            node.codegen_into(code)
+
+        code.splice(
+            f"{self.var}_tile_ptrs = {self.arg_names[list_index]} + ({index_expr})"
+        )
+
+    def index_expr(self, list_index: int):
+        expr = self.indexers[list_index](self.index_vars[list_index])
+        return V.graph.sizevars.simplify_with_ranges(
+            expr, self.range_trees[list_index].var_ranges
+        )
+
+
+class ForeachKernel(Kernel):
+    overrides = TritonOverrides
+
+    @staticmethod
+    def create_kernels(foreach_node):
+        """Creates one or more ForeachKernels if the number of args exceeds CUDA limits."""
+        node_schedule = foreach_node.get_nodes()
+        assert len(node_schedule) >= 1
+        layouts = node_schedule[0].node.get_layouts()
+        list_length = len(layouts)
+        num_lists = sum([node.node.list_arg_count() for node in node_schedule])
+
+        MAX_NUM_ARGS = 370  # number where I would no longer get triton errors
```

Comment:
I don't have access to smaller cards, so I'm not sure this will work, I just went with 4KB arg limit and went smaller and smaller until it worked. Do you have suggestions on how I could determine this statically? I could go smaller like 128 or something.

Reply:
The number of args per kernel doesn't depend on the consumer/non-consumer card, should be fine.